src/goto-programs/remove_const_function_pointers.cpp

```diff
@@ -0,0 +1,1012 @@
+   base_expression - The function call through a function pointer
+   ns - The namespace to use to resolve types
+   symbol_table - The symbol table to look up symbols in
+
+ Outputs:
+
+ Purpose: To take a function call on a function pointer, and if possible
+          resolve it to a small collection of possible values.
+
+\*******************************************************************/
+
+remove_const_function_pointerst::remove_const_function_pointerst(
+  message_handlert &message_handler,
+  const exprt &base_expression,
+  const namespacet &ns,
+  const symbol_tablet &symbol_table):
+    messaget(message_handler),
+    original_expression(base_expression),
+    ns(ns),
+    symbol_table(symbol_table)
+{}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::operator()
+
+  Inputs:
+   out_functions - The functions that (symbols of type ID_code)
+
+ Outputs: Returns true if it was able to resolve the call, false if not.
+          If it returns true, out_functions will be populated by all the
+          possible values the funciton pointer could be.
+
+ Purpose: To take a function call on a function pointer, and if possible
+          resolve it to a small collection of possible values. It will
+          resolve function pointers that are const and:
+          - assigned directly to a function
+          - assigned to a value in an array of functions
+          - assigned to a const struct component
+          Or variations within.
+
+\*******************************************************************/
+
+bool remove_const_function_pointerst::operator()(
+  functionst &out_functions)
+{
+  // Replace all const symbols with there values
+  exprt non_symbol_expression=replace_const_symbols(original_expression);
+  return try_resolve_function_call(non_symbol_expression, out_functions);
+}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::replace_const_symbols
+
+  Inputs:
+   expression - The expression to resolve symbols in
+
+ Outputs: Returns the a modified version of the expression, with all
+          const symbols resovled to their actual values.
+
+ Purpose: To collapse the symbols down to their values where possible
+          This takes a very general approach, recreating the expr tree
+          exactly as it was and ignoring what type of expressions are found
+          and instead recursses over all the operands.
+
+\*******************************************************************/
+
+exprt remove_const_function_pointerst::replace_const_symbols(
+  const exprt &expression) const
+{
+  if(expression.id()==ID_symbol)
+  {
+    if(is_const_expression(expression))
+    {
+      const symbolt &symbol=
+        symbol_table.lookup(expression.get(ID_identifier));
+      if(symbol.type.id()!=ID_code)
+      {
+        const exprt &symbol_value=symbol.value;
```

Comment:
So the answer here is we don't. No attempt is made to validate that const-ness has been respected. 

I initially believed that we didn't need to worry at all about UB and that if the code was doing something illegal, it was acceptable to not worry about it. However, since discussion about null pointers (#524), I am less certain. 

Ultimately, I don't have enough knowledge to make a judgement on this. I was under the impression that we can't easily validate the const-ness before first resolving the FPs resulting in a chicken/egg problem. Unfortunately @martin-cs is away this week, perhaps @danpoe knows more?